jupyter_client/multikernelmanager.py

```diff
@@ -123,6 +130,56 @@ def __del__(self):
 
     def list_kernel_ids(self) -> t.List[str]:
         """Return a list of the kernel ids of the active kernels."""
+        if self.external_connection_dir is not None:
+            external_connection_dir = Path(self.external_connection_dir)
+            if external_connection_dir.is_dir():
+                connection_files = [p for p in external_connection_dir.iterdir() if p.is_file()]
+
+                # remove kernels (whose connection file has disappeared) from our list
+                k = list(self.kernel_id_to_connection_file.keys())
+                v = list(self.kernel_id_to_connection_file.values())
+                for connection_file in list(self.kernel_id_to_connection_file.values()):
+                    if connection_file not in connection_files:
+                        kernel_id = k[v.index(connection_file)]
+                        del self.kernel_id_to_connection_file[kernel_id]
+                        del self._kernels[kernel_id]
+
+                # add kernels (whose connection file appeared) to our list
+                for connection_file in connection_files:
+                    if connection_file in self.kernel_id_to_connection_file.values():
+                        continue
+                    try:
+                        connection_info = json.loads(connection_file.read_text())
+                    except Exception:  # noqa: S112
+                        continue
+                    if not ("kernel_name" in connection_info and "key" in connection_info):
+                        continue
+                    # it looks like a connection file
+                    kernel_id = self.new_kernel_id()
+                    self.kernel_id_to_connection_file[kernel_id] = connection_file
+                    km = self.kernel_manager_factory(
+                        parent=self,
+                        log=self.log,
+                        owns_kernel=False,
+                    )
+                    km.last_activity = utcnow()
+                    km.execution_state = "idle"
+                    km.connections = 1
+                    km.kernel_id = kernel_id
+                    km.shell_port = connection_info["shell_port"]
+                    km.iopub_port = connection_info["iopub_port"]
+                    km.stdin_port = connection_info["stdin_port"]
+                    km.control_port = connection_info["control_port"]
+                    km.hb_port = connection_info["hb_port"]
+                    km.ip = connection_info["ip"]
+                    km.transport = connection_info["transport"]
+                    km.session.key = connection_info["key"].encode()
+                    km.session.signature_scheme = connection_info["signature_scheme"]
+                    km.kernel_name = connection_info["kernel_name"]
```

Comment:
Could we leverage the [`load_connection_info()`](https://github.com/jupyter/jupyter_client/blob/508982f11a3175089b572a0c8277542d9fbd0729/jupyter_client/connect.py#L530) method on the newly-created kernel manager instance?

Reply:
Good point, thanks!